File: dotnet/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example1/Aircraft.cs
namespace MediatorDesignPattern.Example1;

/// <summary>
/// Colleague class - Aircraft object
/// Each aircraft communicates through the mediator (tower), not directly with other aircraft
/// </summary>
public abstract class Aircraft
{
    protected IMediator Mediator;
    public string CallSign { get; }
    public string AircraftType { get; }

    protected Aircraft(IMediator mediator, string callSign, string aircraftType)
    {
        Mediator = mediator;
        CallSign = callSign;
        AircraftType = aircraftType;
    }

    public abstract void Send(string message);
    public abstract void Receive(string message);
    public abstract void RequestLanding();
    public abstract void RequestTakeoff();
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example1/src/Aircraft.ts
import { IMediator } from './IMediator';

export abstract class Aircraft {
  protected mediator: IMediator;
  readonly callSign: string;
  readonly aircraftType: string;

  constructor(mediator: IMediator, callSign: string, aircraftType: string) {
    this.mediator = mediator;
    this.callSign = callSign;
    this.aircraftType = aircraftType;
  }

  abstract send(message: string): void;
  abstract receive(message: string): void;
  abstract requestLanding(): void;
  abstract requestTakeoff(): void;
}
